### PR TITLE
Fix LIBDL TPL/Test issues with Trilinos

### PR DESCRIPTION
**Pull request by jjwilke**
* Fixes issues where KOKKOS_ADD_TEST_LIBRARY creates compile errors in downstream projects
* Defines TPL macros for Trilinos to enable profiling hooks for Trilinos builds
* Activates profile tool tests within Trilinos
* Cleans up KOKKOS_ADD_TEST_LIBRARY

### CODE REVIEW

**Comment by jjwilke** [comment]
This was patched in Trilinos here: https://github.com/trilinos/Trilinos/pull/8233

Need to mirror fixes to Kokkos repo

**Comment by jjwilke** [comment]
This unfortunately needs to be merged for the next release since it is necessary for Trilinos promotion

**Review by ndellingwood — APPROVED** [review]
lgtm, already tested on the Trilinos side

File: core/unit_test/CMakeLists.txt [hunk 1]
```diff
@@ -554,13 +554,12 @@ KOKKOS_ADD_ADVANCED_TEST( UnitTest_PushFinalizeHook_terminate
   endif()
   if(KOKKOS_ENABLE_LIBDL)
 
-    add_library(
+    KOKKOS_ADD_TEST_LIBRARY(
       kokkosprinter-tool SHARED
-      tools/printing-tool.cpp
+      SOURCES tools/printing-tool.cpp
+      DEPLIBS kokkoscore
```
Comment (DavidPoliakoff) [inline core/unit_test/CMakeLists.txt:560]: What does DEPLIBS mean here? It doesn't depend on kokkoscore

**Review by masterleinad — COMMENTED** [review]
Apart from @DavidPoliakoff's question (and the failing build which looks related), this is fine with me.
The only real change for non-Trilinos builds is the ability to choose if a test library should be `static` or `shared` and adding a dependent library.

**Comment by jjwilke** [comment]
@masterleinad where is the error? I can't seem to find it.

File: core/unit_test/CMakeLists.txt (hunk 1 above)
Comment (jjwilke) [reply core/unit_test/CMakeLists.txt:560]: I can remove it.

Comment (jjwilke) [reply core/unit_test/CMakeLists.txt:560]: But, yeah, it's a way to make test libraries depend on kokkoscore, etc

**Comment by crtrott** [comment]
You have to go to "Blue Ocean" on the jenkins site.

File: core/unit_test/CMakeLists.txt [hunk 2]
```diff
@@ -554,13 +554,11 @@ KOKKOS_ADD_ADVANCED_TEST( UnitTest_PushFinalizeHook_terminate
   endif()
   if(KOKKOS_ENABLE_LIBDL)
 
-    add_library(
+    KOKKOS_ADD_TEST_LIBRARY(
```
[the remark this answers is not in the document]
Comment (masterleinad) [reply core/unit_test/CMakeLists.txt:557]: I am fine with this pull request if you just revert the changes to this file.

**Review by masterleinad — CHANGES_REQUESTED** [review]
Please address https://github.com/kokkos/kokkos/pull/3543/files#r514421233.

**Pull request merged**